### PR TITLE
feat: Exponential sentry error suppression #385

### PR DESCRIPTION
**Pull request by pritamghanghas**
* added function to be used as before_send hook with sentry sdk

**[Issue](https://github.com/NebraLtd/hm-diag/issues/385)**

- Link: https://github.com/NebraLtd/hm-diag/issues/385
- Summary: add support for exponential suppression of error messages by using before_send sentry hook

**How**

**Screenshots**

**References**

**Checklist**

- [x] Tests added
- [x] Cleaned up commit history (rebase!)
- [ ] Documentation added
- [x] Thought about variable and method names

### CODE REVIEW

File: hm_pyhelper/tests/data/sentry/sample_exception_event.json [hunk 1]
```diff
@@ -0,0 +1,130 @@
+{
```
Comment (pritamghanghas) [inline hm_pyhelper/tests/data/sentry/sample_exception_event.json:1]: Ignore the json files. They are just sample dumps sentry events to serve as test data.

File: hm_pyhelper/util/sentry.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    # with a truncated string as well
+    fingerprint_bytes = str(value_to_fingerprint).encode('utf-8')
+    sha256_hash = hashlib.sha256()
+    sha256_hash.update(fingerprint_bytes)
+    fingerprint = sha256_hash.hexdigest()
+
+    return fingerprint
+
+
+def process_fingerprint(fingerprint: str) -> bool:
```
Comment (pritamghanghas) [inline hm_pyhelper/util/sentry.py:36]: we don't account for elapsed time for suppression. Didn't feel the need for it.

**Review by MuratUrsavas — APPROVED** [review]

File: hm_pyhelper/util/sentry.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+# maximum suppression interval. It grows exponentially till this value is reached.
+max_suppression_count = 128
+
+
+def sentry_fingerprint(hints: Dict) -> Union[str, None]:
+    """ return sha256 of log_record/exc_info as fingerprint"""
+    fingerprint_info_key = 'log_record'
+    if 'exc_info' in hints:
+        fingerprint_info_key = 'exc_info'
```
Comment (kashifpk) [inline hm_pyhelper/util/sentry.py:18]: @pritamghanghas - This seems to imply that same exception/error for different values will generate different finger prints? For example a `ValueError` for an expected number receiving values `not a number` or `really not a number` would create 2 different fingerprints right? as the resulting exception's info will contain these values too.

I think this is reasonable for now though as having the exc_info with values would give more insight into the error.

File: hm_pyhelper/util/sentry.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+def process_fingerprint(fingerprint: str) -> bool:
+    """
+    does book keeping based on event fingerprint.
+    returns true if the event should be allowed to propagate
+    """
+    # decide whether to send or suppress
+    send_event = False
+    if fingerprint in event_counters:
+        event_count = event_counters[fingerprint][EVENT_COUNT_KEY]
+        suppression_count = event_counters[fingerprint][SUPPRESSION_COUNT_KEY]
```
Comment (kashifpk) [inline hm_pyhelper/util/sentry.py:45]: For counters using a [Counter](https://docs.python.org/3/library/collections.html#collections.Counter) is the preferred way as it is thread-safe.

File: hm_pyhelper/util/sentry.py (hunk 3 above)
Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py:18]: @kashifpk we don't fingerprint exception's. I didn't see a point in trying to suppress uncaught exception. This is only for logging.error() with or without exc_info. I agree with different parameter value resulting in different fingerprint.  I didn't think of it. 

I can remove exc_info as log_record will also have it. But to make fingerprint cover "value differing cases", I think I should use only file and line number from LogRecord for better suppresion.
eg. fingerprint = f"{LogRecord.filename}:{LogRecord.lineno}:{LogRecord.funcName}"
basically not include anything about actual values in the fingerprint calculation.

File: hm_pyhelper/util/sentry.py (hunk 4 above)
Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py:45]: sure, will make that change.

File: hm_pyhelper/util/sentry.py (hunk 3 above)
Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py:18]: sentry also collects differing values together. Initially I had looked for a public api to get fingerprint from sentry itself. Didn't find it.

Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py]: update code to follow proposal

File: hm_pyhelper/util/sentry.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def sentry_fingerprint(hints: Dict) -> Union[str, None]:
+    """ return sha256 of log_record as fingerprint"""
+
+    if 'log_record' not in hints:
+        return None
+
+    log_record = hints.get('log_record', None)
+    fingerprint = f"{log_record.filename}:{log_record.lineno}:{log_record.funcName}"
```
Comment (marvinmarnold) [inline hm_pyhelper/util/sentry.py:21]: nit: funcname for consistency

Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py:21]: not our data structure, it comes from Logging module.

File: hm_pyhelper/util/sentry.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        event_counters[fingerprint] = Counter({
+            EVENT_COUNT_KEY: 1,
+            SUPPRESSION_COUNT_KEY:  1
+            })
+        send_event = True
+
+    return send_event
+
+
+def before_send_filter(event: Dict, hints: Dict) -> Union[Dict, None]:
```
[the remark this answers is not in the document]
Comment (pritamghanghas) [reply hm_pyhelper/util/sentry.py:53]: comes from sentry callback, I am adding comments to clear it up.
